acceptance: 6-hour continuous block_writer test

cuongdo opened the pull request:
This test uses Terraform to create a 3-node test cluster and generates 6
hours of continuous block_writer load against it.

cc @tschottdorf @petermattis

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7973)

<!-- Reviewable:end -->

Commit:
acceptance: 6-hour continuous block_writer test

This test uses Terraform to create a 3-node test cluster and generates 6
hours of continuous block_writer load against it.

tschottdorf commented on acceptance/terrafarm/farmer.go:
```diff
@@ -203,6 +203,16 @@ func (f *Farmer) Exec(i int, cmd string) error {
+// ExecWriter executes the given command on the i-th writer, returning
+// (in that order) stdout, stderr and an error.
+func (f *Farmer) ExecWriter(i int, cmd string) error {
+	stdout, stderr, err := f.ssh(f.Writers()[i], f.defaultKeyFile(), cmd)
```
afaics the only difference here is `s/Writers/Nodes/`, so I think code sharing would make sense.

tschottdorf commented on acceptance/allocator_terraform/load.tf:
```diff
@@ -4,6 +4,10 @@ variable "example_block_writer_instances" {
 }
 
+variable "load_duration" {
+  default = "0" # unlimited
```
Why do you need this? Appears to me that the duration is really governed by the test runner (assuming you can be off by a few minutes). You could make the tests omit the duration completely and I don't see what would be lost.

cuongdo replied:
`block_writer` needs to terminate and print benchmark stats as I made it do in https://github.com/cockroachdb/examples-go/pull/72 (currently blocked because of `grpc-go`-related build issue).

I guess I could make the stats print in a `defer` block in `block_writer` so that it always shows up.

tschottdorf replied:
Yeah, it *should* be possible to always print the stats when block writer exists (short of being `SIGKILL`ed). I think that would be preferable, especially if it avoids the double-timeout handling and awkward terrafrorm plumbing here.

tschottdorf commented on acceptance/continuous_load_test.go:
```diff
@@ -0,0 +1,145 @@
+	}
+}
+
+func TestBlockWriter6Hours(t *testing.T) {
```
How about `TestContinuousLoad_3Nodes` or `TestContinuousLoad_3NodesBlockWriter`(and besides, why can't the number of nodes be passed in from the outside? I see that that doesn't make sense for all of the allocator tests, but this one starts from scratch).
I think it also makes sense to avoid the three node case (maybe four or five?) because that never has to move replicas.

tschottdorf commented on acceptance/terrafarm/farmer.go:
```diff
@@ -271,6 +281,26 @@ func (f *Farmer) Assert(t *testing.T) {
 
+// AssertStates verifies that the cluster state matches the provided map
+// of process names (e.g. cockroach) to expected states (e.g. RUNNING).
+func (f *Farmer) AssertStates(t *testing.T, procToExpState map[string]string) {
```
You could also lose that map with my suggestion - simply check that everything is up everywhere.

cuongdo replied:
What if something is expected to be down? After the load test period is over, `block_writer` should exit normally.

tschottdorf replied:
Might make sense eventually, but you're only allowing every process to be the same on each of the nodes, so it's not very expressive. It seems to be useful pretty much only for your specific use case, so unless you need to keep it for that I'd just bring it back when the need arises. I thought briefly on when that would happen, and couldn't see it right on the horizon.

tschottdorf commented:
LGTM

tschottdorf commented on acceptance/continuous_load_test.go:
```diff
@@ -0,0 +1,145 @@
+
+	// Ensure that the test timeout allows the load test to run and tear down
+	// properly, to avoid Terraform resource leaks.
+	const destroyTime = 10 * time.Minute
```
Could you make this more general? We want all of the tests to (be able to) use this. Perhaps ```go
ctx := context.WithTestTimeout(parent context.Context) // has deadline set at (testduration-5min) (I think it's fair to just hard-code 5min for now).
```

tschottdorf commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'll defer to Tobias on the specifics of the change. Excited to see this go in.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-:-KNI-I8SVgvt-V0MPqN1:-452077244)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/continuous_load_test.go, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-KNFYCjYNLJ6TLCWZhpZ-r1-69:-KNYfohjmudTBSd7XOcT:1029143288) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e9b1a842e9fcc010399baa5b70172cb3bdb2932/acceptance/continuous_load_test.go#L69)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Could you make this more general? We want all of the tests to (be able to) use this. Perhaps 
> 
> 
> 
> ```go
> 
> ctx :=WithTestTimeout(parent context.Context)
> 
> // ctx has deadline set at (testduration-5min)
> 
> // (I think it's fair to just hard-code 5min for now).
> 
> ``` 
</details>
added `WithTeardownTimeout` in `util_test.go`

---

*[acceptance/continuous_load_test.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-KNFYCjYNLJ6TLCWZhpZ-r1-137:-KNYflfQxWQLA47OOFFs:648872496) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e9b1a842e9fcc010399baa5b70172cb3bdb2932/acceptance/continuous_load_test.go#L137)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> How about `TestContinuousLoad_3Nodes` or `TestContinuousLoad_3NodesBlockWriter`(and besides, why can't the number of nodes be passed in from the outside? I see that that doesn't make sense for all of the allocator tests, but this one starts from scratch).
> 
> I think it also makes sense to avoid the three node case (maybe four or five?) because that never has to move replicas.
</details>
Done. Now I just use the existing `-nodes` flag. `BlockWriter` needs to be part of the name, because there will soon be a corresponding  test that uses `photos`

---

*[acceptance/terrafarm/farmer.go, line 209 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-KNFYCjYNLJ6TLCWZhpY-r1-209:-KNXydU_l-UsNPS3_SaY:-327604333) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e9b1a842e9fcc010399baa5b70172cb3bdb2932/acceptance/terrafarm/farmer.go#L209)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> afaics the only difference here is `s/Writers/Nodes/`, so I think code sharing would make sense.
</details>
This is no longer needed after some other changes, so I've removed this method.

---

*[acceptance/terrafarm/farmer.go, line 286 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-KNFYCjYNLJ6TLCWZhpY-r1-286:-KNXz2mEMtgJhHVPKy9C:-1315767301) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e9b1a842e9fcc010399baa5b70172cb3bdb2932/acceptance/terrafarm/farmer.go#L286)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Might make sense eventually, but you're only allowing every process to be the same on each of the nodes, so it's not very expressive. It seems to be useful pretty much only for your specific use case, so unless you need to keep it for that I'd just bring it back when the need arises. I thought briefly on when that would happen, and couldn't see it right on the horizon.
</details>
Done. No longer needed with other changes.

---

*[acceptance/allocator_terraform/load.tf, line 8 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-KNFYCjYNLJ6TLCWZhpU-r1-8:-KNXTRVIRYhbIyW-gYsa:-1189714655) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e9b1a842e9fcc010399baa5b70172cb3bdb2932/acceptance/allocator_terraform/load.tf#L8)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yeah, it *should* be possible to always print the stats when block writer exists (short of being `SIGKILL`ed). I think that would be preferable, especially if it avoids the double-timeout handling and awkward terrafrorm plumbing here.
</details>
Done (in `examples-go`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Rebased and addressed Tobi's comments

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-:-KNYquo6RjlVRufInd5F:1757931691)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[acceptance/continuous_load_test.go, line 123 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-KNYuVdDHsT-di6Na-4H:-KNYuVdDHsT-di6Na-4I:-639362133) ([raw file](https://github.com/cockroachdb/cockroach/blob/8622f9899e60d8877650e272dd15b0187c621f24/acceptance/continuous_load_test.go#L123)):*
> ```Go
> 		timeout, err := time.ParseDuration(fl.Value.String())
> 		if err == nil {
> 			prefix += timeout.String()
> ```

Does this print as `6h0m0s` and if so, do you want to massage it as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-:-KNYulLMRd3z02My9TI-:1417088721)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Fixed a `make check` error. Also, we now pass through a stringified `test.timeout` to `block_writer` to use as the benchmark name in the benchmark output, because `block_writer` doesn't know how long we *intended* to run it for.

There are still occasional test failures that I haven't been able to track down. It manifests as a test timeout with no intervening health health checks. It looks like we're freezing on a `(*Farmer).Assert()` somewhere. It seems like there are sometimes "stuck" nodes, where even SSH isn't responsive. I haven't been able to diagnose, but I think the test is stable enough to integrate with TeamCity.

---

Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[acceptance/continuous_load_test.go, line 123 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-KNYuVdDHsT-di6Na-4H:-KNbqPYTuLRKM6fxkCDO:-1304039438) ([raw file](https://github.com/cockroachdb/cockroach/blob/8622f9899e60d8877650e272dd15b0187c621f24/acceptance/continuous_load_test.go#L123)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Does this print as `6h0m0s` and if so, do you want to massage it as well?
</details>
Yes, I've moved the logic to `continuousLoadtest.shortTestTimeout`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7973#-:-KNbqXBFy3IyWQoNffyI:827459130)*
<!-- Sent from Reviewable.io -->
